File: src/fuzzy/rule/abstract_rule.py
from abc import ABC, abstractmethod


class AbstractRule(ABC):
    _antecedent = None
    _consequent = None

    def __init__(self, antecedent, consequent):
        self._antecedent = antecedent
        self._consequent = consequent

    def get_antecedent(self):
        return self._antecedent

    def get_consequent(self):
        return self._consequent

    def get_compatible_grade(self, attribute_vector):
        return self._antecedent.get_compatible_grade(attribute_vector)

    def get_compatible_grade_value(self, attribute_vector):
        return self._antecedent.get_compatible_grade_value(attribute_vector)

    def get_class_label(self):
        return self._consequent.get_class_label()

    def get_class_label_value(self):
        return self._consequent.get_class_label_value()

    def equals_class_label(self, other):
        return self._consequent().get_class_label() == other.get_consequent().get_class_label()

    def is_rejected_class_label(self):
        return self._consequent.get_class_label().is_rejected()

    def get_rule_weight(self):
        return self._consequent.get_rule_weight()

    def get_rule_weight_value(self):
        return self._consequent.get_rule_weight_value()

    def set_rule_weight_value(self, rule_weight_value):
        self._consequent.set_rule_weight_value(rule_weight_value)

    def set_class_label_value(self, class_label_value):
        self.get_consequent().set_class_label_value(class_label_value)

    @abstractmethod
    def get_fitness_value(self, attribute_vector):
        pass

    # TODO: Add rule builder ?

File: src/mofgbmlpy/core/fuzzy/rule/abstract_rule.cpp
//
// Created by Robin on 16/07/2025.
//

#include "abstract_rule.hpp"

AbstractRule::AbstractRule(Antecedent* antecedent, AbstractConsequent* consequent)
    : antecedent(antecedent), consequent(consequent)
{
    if (!this->antecedent || !this->consequent) {
        throw std::invalid_argument("Antecedent and consequent must not be null.");
    }
}

AbstractRule::AbstractRule(const AbstractRule& other)
{
    if (!other.antecedent || !other.consequent) {
        throw std::invalid_argument("Antecedent and consequent must not be null.");
    }

    Antecedent* antecedent_copy = other.antecedent->clone();
    this->antecedent = antecedent_copy;

    AbstractConsequent* consequent_copy = other.consequent->clone();
    this->consequent = consequent_copy;
}

Antecedent* AbstractRule::get_antecedent() const
{
    return antecedent;
}

AbstractConsequent* AbstractRule::get_consequent() const
{
    return consequent;
}

void AbstractRule::set_consequent(AbstractConsequent* new_consequent)
{
    this->consequent = new_consequent;
}

std::vector<double> AbstractRule::get_membership_values(const std::vector<double>& attribute_vector) const
{
    return antecedent->get_membership_values(attribute_vector);
}

double AbstractRule::get_compatible_grade_value(const std::vector<double>& attribute_vector) const
{
    return antecedent->get_compatible_grade_value(attribute_vector);
}

AbstractClassLabel* AbstractRule::get_class_label() const
{
    return consequent->get_class_label();
}

bool AbstractRule::is_rejected_class_label() const
{
    return consequent->get_class_label()->is_rejected();
}

AbstractRuleWeight* AbstractRule::get_rule_weight() const
{
    return consequent->get_rule_weight();
}

int AbstractRule::get_length() const
{
    return antecedent->get_length();
}

Knowledge* AbstractRule::get_knowledge() const
{
    return antecedent->get_knowledge();
}

FuzzySet* AbstractRule::get_fuzzy_set_object(int dim_index) const
{
    const int fuzzy_set_index = antecedent->get_antecedent_indices()[dim_index];
    return get_knowledge()->get_fuzzy_set(dim_index, fuzzy_set_index);
}

int AbstractRule::get_antecedent_array_size() const
{
    return antecedent->get_array_size();
}

std::string AbstractRule::get_var_name(int dim_index) const
{
    return get_knowledge()->get_fuzzy_variable(dim_index)->get_name();
}

std::string AbstractRule::get_linguistic_representation() const
{
    return "IF\t" + antecedent->get_linguistic_representation() +
           " THEN " + consequent->get_linguistic_representation() +
           " RW: " + std::string(*consequent->get_rule_weight());
}

AbstractRule::operator std::string() const
{
    return "Antecedent: " + std::string(*antecedent) + " => Consequent: " + std::string(*consequent);
}

bool AbstractRule::operator==(const AbstractRule& other) const
{
    return *antecedent == *other.antecedent && *consequent == *other.consequent;
}

std::string AbstractRule::to_string() const
{
    return static_cast<std::string>(*this);
}

